src/platform_impl/macos/window.rs

```diff
@@ -149,6 +151,7 @@ pub struct SharedState {
     /// bar in exclusive fullscreen but want to restore the original options when
     /// transitioning back to borderless fullscreen.
     save_presentation_opts: Option<NSApplicationPresentationOptions>,
+    accepts_first_mouse: bool,
```

Comment:
Actually, since we're accessing this from inside `WinitView`, I would really prefer if this variable could live in `WinitView`.

I know that may be a bit difficult, you'll have to change the initializer to pass it through that - please do tell if you need help with it!

(I'm trying to move away from having one big `SharedState`/`ViewState` struct, and instead just have normal ivars on the objects).

Reply:
> please do tell if you need help with it!

Yes please :)

First of all, what should the variable definition look like? `accepts_first_mouse: IvarDrop<Box<bool>>` perhaps? Is this the right place? https://github.com/rust-windowing/winit/blob/ba49db2cb9932f2ec90ba9466f980b19bb102081/src/platform_impl/macos/view.rs#L135-L139

Then, this `new` function should take an extra `accepts_first_mouse: bool` parameter?
https://github.com/rust-windowing/winit/blob/ba49db2cb9932f2ec90ba9466f980b19bb102081/src/platform_impl/macos/view.rs#L916-L918

It returns a `Id<Self, Shared>`, and I don't understand what that means, but I guess it's a pointer to the new instance; should I store it in a variable, set `accepts_first_mouse` on it, and return it?

Alternatively, would it make sense to add a `set_accepts_first_mouse` method, to be called after `new`? I can't get this to work though; seems it's not possible to get a mutable reference to the view class:

```
error[E0596]: cannot borrow data in dereference of `Id<WinitView, Shared>` as mutable
   --> src/platform_impl/macos/window.rs:361:9
    |
361 |         view.set_accepts_first_mouse(pl_attrs.accepts_first_mouse);
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot borrow as mutable
    |
    = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `Id<WinitView, Shared>`
```

---
BTW, the pull request has _Allow edits and access to secrets by maintainers_ checked, so you should be able to push to it, if you want.

Reply:
> First of all, what should the variable definition look like? `accepts_first_mouse: IvarDrop<Box<bool>>` perhaps?

`accepts_first_mouse: bool` would suffice (the `IvarDrop` is only used for types like `Box` and `Id`, while `bool` doesn't need to be dropped (and implements `Encode`)).

> Is this the right place?
> 
> https://github.com/rust-windowing/winit/blob/ba49db2cb9932f2ec90ba9466f980b19bb102081/src/platform_impl/macos/view.rs#L135-L139

Yup.

> Then, this `new` function should take an extra `accepts_first_mouse: bool` parameter?
> 
> https://github.com/rust-windowing/winit/blob/ba49db2cb9932f2ec90ba9466f980b19bb102081/src/platform_impl/macos/view.rs#L916-L918

Indeed!

> It returns a `Id<Self, Shared>`, and I don't understand what that means, but I guess it's a pointer to the new instance

It's a pointer to the new instance, that will automatically `release` on `Drop`. Think of `Id<T, Shared>` as the Objective-C equivalent of `Arc`, while `Id<T, Owned>` is `Box`.

> should I store it in a variable, set `accepts_first_mouse` on it, and return it?

You should change the `initWithId:` selector (that is just an arbitary name I picked) such that it can accept another argument. For example `initWithWindow:acceptsFirstMouse:`.

> Alternatively, would it make sense to add a `set_accepts_first_mouse` method, to be called after `new`? I can't get this to work though; seems it's not possible to get a mutable reference to the view class:
> 
> ```
> error[E0596]: cannot borrow data in dereference of `Id<WinitView, Shared>` as mutable
>    --> src/platform_impl/macos/window.rs:361:9
>     |
> 361 |         view.set_accepts_first_mouse(pl_attrs.accepts_first_mouse);
>     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot borrow as mutable
>     |
>     = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `Id<WinitView, Shared>`
> ```

It could work if you returned `Id<Self, Owned>` above. I'm in the progress of figuring out what we should actually do here, leaning towards having e.g. `accepts_first_mouse: Cell<bool>` if we'd want it to be mutable. See https://github.com/madsmtm/objc2/issues/265 for details.

> BTW, the pull request has _Allow edits and access to secrets by maintainers_ checked, so you should be able to push to it, if you want.

Huh, I tried it before and it didn't work, but maybe I messed something up. If you want to have a go at doing stuff yourself, feel free to, otherwise I can do it (in a few days probably)

Reply:
The `initWithWindow:acceptsFirstMouse:` selector feels a bit janky, and I guess `accepts_first_mouse` _should_ be mutable, but TBH I'm not up to messing with Rust <-> ObjC much more :)

Reply:
Yeah, I agree that `accepts_first_mouse` should be mutable, again, working on this in https://github.com/madsmtm/objc2/issues/265 ;)